src/main/java/fi/aalto/cs/apluscourses/model/Exercise.java

```diff
@@ -28,6 +28,8 @@
 
   private final boolean submittable;
 
+  private boolean inGrading;
```

Comment:
I realized that if you merge my PR, then this should be either `volatile` or `AtomicBoolean`, because in my PR the boolean gets modified in the thread of `SubmissionStatusUpdater` and then read in other threads when the tree is updated. Although atomicity isn't strictly necessary, I suggest going with `AtomicBoolean`, since the `submissionResults` variable in `Exercise` also uses a synchronized list (unnecessarily).